### ISSUE TITLE
Add Hashicorp Vault secrets provider (direct, not transit)

### ISSUE THREAD

**Issue by nesl247** [pulumi/pulumi#5211]
We'd like to see pulumi make use of hashicorp vault for retrieving secrets, but not using the transit engine. We want all of the values to be stored in the vault, and for pulumi to simply fetch the values at runtime.

The use case for this is that we use the values in multiple places, and don't want to have to duplicate values between systems.

**Comment by lukehoban** [comment]
Note that it should be possible to do this already today - by writing code using the [Vault client libraries](https://www.vaultproject.io/api/libraries.html) to connect to and pull secrets out of vault and use as inputs to your resources.

Do you actually *want* the config settings to be set via `pulumi config` (and thus at least the keys duplicated with your Vault)?  We could in principle support Pulumi config that "redirects" to secrets stored in specific secrets stores.  This is something we've considered in the past - but not yet pursued due to the workaround being in principle "better" in some ways.  Would love to understand whether using Vault APIs works and if not what exactly the pain points are that we could make better with a more integrated solution.

**Comment by nesl247** [comment]
Yes, we want the configs to be set via `pulumi config`. With this, we can use `requireSecret()`, unlike doing it as you mentioned (which we already do), and get the values properly marked as secrets.
